sgkit/stats/grm.py

```diff
@@ -147,3 +148,522 @@ def genomic_relationship(
         }
     )
     return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def _expand_matrix(m, like, x_indices, y_indices):
+    _M = da.zeros((like.shape[0], m.shape[1]), chunks=(like.chunks[0], m.chunks[1]))
+    _M[x_indices] = m
+    M = da.zeros_like(like)
+    M[:, y_indices] = _M
+    return M
+
+
+def _hybrid_full_matrix_Martini(Ainv, G, tau=1, omega=1):
+    # calculate sub-matrices of H
+    if (tau == 1) and (omega == 1):
+        H = G
+    elif omega == 1:
+        H = 1 / tau * G
+        # H = da.linalg.inv(tau * da.linalg.inv(G))
+    else:
+        H = da.linalg.inv(tau * da.linalg.inv(G) + (1 - omega) * Ainv)
+    return H
+
+
+def _hybrid_sub_matrix_Martini(A, G, Gidx, tau=1, omega=1, A22inv=None):
+    n = len(A)
+    # indices of non G-matrix
+    bools = np.ones(n, bool)
```

Comment:
There are a few points in this function where `np` methods are used - I'm not familiar enough with the method to know if it is the right thing to do to bring computation out of dask at those points, but a comment explaining why it's ok (for example where it is known the arrays will be small at that point) would be good.

Reply:
Good call, these arrays are used to slice the A-matrix into sub-matrices. They need to be evaluated eagerly so that the sub-matrices have known size. They are all 1-d arrays with maximum size of n_samples